solutions/usaco-1015.mdx

```diff
@@ -0,0 +1,234 @@
+---
+
+[Original Problem](http://www.usaco.org/index.php?page=viewproblem2&cpid=1015)
+
+[Official Analysis](http://www.usaco.org/current/data/sol_triangles_silver_feb20.html)
+
+## Implementation
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+#define MAXN 100005
+#define MOD 1000000007
+#define ll long long
+#define mp make_pair
+#define sz(x) (int) (x).size() 
+#define pb push_back
+
+using namespace std;
+
+void setIO(string prob = "") {
+	if (sz(prob)) {
+		freopen((prob + ".in").c_str(), "r", stdin);
+		freopen((prob + ".out").c_str(), "w", stdout);
+	}
+}
+
+struct Fence {
+	// heightsum is the sum of the heights of all the triangles
+	// that use this fence as an anchor point, basesum is the
+	// sum of the bases of all the triangles that use this
+	// fence as an anchor point; recall that the anchor point
+	// of a right triangle is the vertex where the right
+	// angle is.
+	int x, y;
+	ll heightsum, basesum;
+};
+
+// fences are our fences, xcoords represents all of the possible
+// x coordinates of our fences, ycoords represents all of the
+// possible y coordinates of our fences
+Fence fences[MAXN];
+vector<pair<int,int>> xcoords[20001];
+vector<pair<int,int>> ycoords[20001];
+
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	setIO("triangles");
+
+	int n; cin >> n;
+	for (int i = 0; i < n; i++) {
+		cin >> fences[i].x >> fences[i].y;
+		// we add 100000 to make all of our coordinates positive
+		// so that we won't have negative indices
+		xcoords[fences[i].x + 10000].pb(mp(fences[i].y, i));
+		ycoords[fences[i].y + 10000].pb(mp(fences[i].x, i));
+	}
+
+	for (int i = 0; i <= 20000; i++) {
+		if (sz(xcoords[i]) > 0) {
+			// cur is the value of the current s_i
+			ll cur = 0;
+			// first we sort all of the y coordinates of each set
+			// of points with the same x coordinate
+			sort(xcoords[i].begin(), xcoords[i].end());
+			// then we compute the value s_1 of this set:
+			// the sum of the heights of all the triangles that
+			// have anchor point at (i, xcoords[i][0].first)
+			for (int j = 1; j < sz(xcoords[i]); j++) {
+				cur+= xcoords[i][j].first - xcoords[i][0].first;
+			}
+			fences[xcoords[i][0].second].heightsum = cur;
+			// then we compute the rest of the s_i for this set
+			for (int j = 1; j < sz(xcoords[i]); j++) {
+				cur+= (2 * j - sz(xcoords[i])) * (xcoords[i][j].first - xcoords[i][j - 1].first);
+				fences[xcoords[i][j].second].heightsum = cur;
+			}
+		}
+	}
+
+	// we do the sums of bases in exactly the same way
+	for (int i = 0; i <= 20000; i++) {
+		if (sz(ycoords[i]) > 0) {
+			ll cur = 0;
+			sort(ycoords[i].begin(), ycoords[i].end());
+			for (int j = 1; j < sz(ycoords[i]); j++) {
+				cur+= ycoords[i][j].first - ycoords[i][0].first;
+			}
+			fences[ycoords[i][0].second].basesum = cur;
+			for (int j = 1; j < sz(ycoords[i]); j++) {
+				cur+= (2 * j - sz(ycoords[i])) * (ycoords[i][j].first - ycoords[i][j - 1].first);
```

Comment:
is += intentionally spaced like (and the rest of the code)

Reply:
like when you perform arithmetic it's `a+= b` instead of `a += b`

Reply:
oh yeah that's intentional, should i change it?

Reply:
i think it's more standardized so that would be ideal, also because it differs from other arithmetic operators ex. `2 * j` being spaced normally.

Reply:
tl;dr yes

Reply:
i mean idk i used this convention in my video explanations so it might be weird if i changed it now